File: ruby/binary/binary.rb
# Module for conversion between decimal and binary numbers
module Binary
  def self.to_decimal(binary)
    raise ArgumentError, 'Invalid input' if binary.to_s =~ /[^01]/
    binary.to_s.reverse.chars.each_with_index.map do |val, idx|
      val.to_i * 2**idx
    end.inject(:+)
  end

  def self.to_binary(decimal)
    raise ArgumentError, 'Invalid input' if decimal.to_s =~ /[^0-9]/
    decimal = decimal.to_i
    binary_n = []
    until decimal < 1
      binary_n.unshift(decimal % 2)
      decimal /= 2
    end
    binary_n.join.to_i
  end
end

module BookKeeping
  VERSION = 3
end

File: javascript/binary/binary.js
let Binary = function(binary) {
  if (binary.match(/[^01]/)) {
    this.binary = '0';
  } else {
    this.binary = binary;
  }
}

Binary.prototype.toDecimal = function() {
  let array = this.binary.split('').reverse();
  let decimal = array.map((el, i) => {
    return (parseInt(el) * Math.pow(2, i));
  });
  return decimal.reduce((acc, el) => acc + el);
}

module.exports = Binary;